.github/workflows/stackhpc-container-image-build.yml

```diff
@@ -208,57 +209,71 @@ jobs:
-        if: github.event.inputs.scan-push == 'false'
+        run: if [ $(wc -l < image-scan-output/clean-images.txt) -le 0 ]; then exit 1; else cp image-scan-output/clean-images.txt image-scan-output/pushed-images.txt; fi
+        if: ${{ !inputs.push-dirty }}
 
-      - name: Append dirty images to clean list
+      - name: Append dirty images to push list
         run: |
-          cat image-scan-output/dirty-images.txt >> image-scan-output/clean-images.txt
-        if: github.event.inputs.scan-push == 'true'
+          cp image-scan-output/clean-images.txt image-scan-output/pushed-images.txt
+          cat image-scan-output/dirty-images.txt >> image-scan-output/pushed-images.txt
+        if: inputs.push-dirty
 
       - name: Push images
         run: |
+          touch image-scan-output/push-failed-images.txt
           source venvs/kayobe/bin/activate &&
           source src/kayobe-config/kayobe-env --environment ci-builder &&
           kayobe playbook run ${KAYOBE_CONFIG_PATH}/ansible/docker-registry-login.yml &&
 
           while read -r image; do
             # Retries!
-            for i in {1..10}; do 
-              docker push ${image} && break || sleep 5
+            for i in {1..5}; do 
+              if docker push $image; then
+                echo "Pushed $image"
+                break
+              elif $i == 5; then
+                echo "Failed to push $image"
+                echo $image >> image-scan-output/push-failed-images.txt
+              else
+                echo "Failed on retry $i"
+                sleep 5
+              fi;
             done
-          done < image-scan-output/clean-images.txt
+          done < image-scan-output/pushed-images.txt &&
+          mv image-scan-output image-build-logs
         shell: bash
         env:
           KAYOBE_VAULT_PASSWORD: ${{ secrets.KAYOBE_VAULT_PASSWORD }}
-        if: github.event.inputs.push == 'true'
+        if: inputs.push
 
-      - name: Upload pushed container images artifact
+      - name: Upload output artifact
         uses: actions/upload-artifact@v4
         with:
-          name: ${{ matrix.distro }}-pushed-container-images
-          path: image-scan-output/clean-images.txt
+          name: ${{ matrix.distro }}-logs
+          path: image-build-logs
           retention-days: 7
 
       - name: Fail when images failed to build
         run: exit 1
         if: steps.build_overcloud_images.outcome == 'failure' || steps.build_seed_images.outcome == 'failure'
 
+      - name: Fail when images failed to push
+        run: if [ $(wc -l < image-build-logs/push-failed-images.txt) -gt 0 ]; then cat image-build-logs/push-failed-images.txt && exit 1; fi
+        if: always()
+
+      - name: Fail when images failed scanning
+        run: if [ $(wc -l < image-build-logs/dirty-images.txt) -gt 0 ]; then cat image-build-logs/dirty-images.txt && exit 1; fi
+        if: ${{ !inputs.push-dirty }}
```

Comment:
It might be helpful to collapse these into a single step, in case there are multiple types of failure. e.g.

These images failed to build:
These images failed to push:
These images failed their scan:

Reply:
Although perhaps we don't have info to hand on which failed to build?

Reply:
Yeah, I don't think we can really get that list easily. We just build what we can then move on. I added if statements to the second two thinking you could have three distinct steps, and quickly see from the actions log if any of those steps failed.

Reply:
I did add [this thing](https://review.opendev.org/c/openstack/kolla/+/881567) which could be used to get a list of failed images. It would need to be backported though. Possibly something "for the future".

Reply:
Do we have a Jira ticket/doc somewhere with all these grand plans in them?

Reply:
Build logs are now included in the artifact, and the task that fails when images did not build tells operators to look in there